### DISCUSSION TITLE
Hull box improvements

### DISCUSSION THREAD

**Discussion by NotWendy** [FakeFishGames/Barotrauma#15123] [Ideas and suggestions]
(This post is about the blue rectangles which mark rooms. Calling them "hulls" is misleading.)

Here are some suggestions I have for improving the functionality of hull boxes:
* Water % property - a text field which lets you edit the amount of water the hull box starts out with.
`water ` console command is a weird approach and doesn't allow for any precision. Turning it into a text input makes things a lot easier.
* Always flooded - a checkbox. Making alien ruins or a wreck and want a hull box to be undrainable? Turn on this setting. Forces the water level to always be at 100%. (Or maybe forces it to maintain the water level you set in the water % property?)
For bonus optimization points, don't process the flow of water from the open ocean into permanently flooded hull boxes.
* Always oxygenated - a checkbox. Want to make a safe room for NPCs for some kind of event? Enable this property instead of adding a hidden oxygen generator. Forces the oxygen level to always be at maximum.
* General change - water removes oxygen from the hull boxes during a flood. For example, if a hull box is 60% flooded, its max oxygen level is only 40%. No more "100% flooded, 100% oxygenated" hull boxes. We're not fish, we don't care how much oxygen is dissolved in water.
For the sake of performance, we probably shouldn't make water push oxygen around the submarine, just straight up delete it from the hull box. Assume it bubbled out into the open ocean.
This makes repairing the oxygen generator after a flood more important, and Jove knows the oxygen generator needs to actually feel useful. At the moment, breathing and fire are the only two things which can remove oxygen from the submarine.

I asked around in the Discord server and people seem to agree with these. I think the more convenient water editing and the change to how flooding affects the oxygen level are the most important parts of the suggestion.

**Comment by Regalis11** [comment]
While I agree with the other suggestions, I somewhat disagree with the idea of making water remove oxygen from the hulls. 

The way I see it, the oxygen value in a flooded hull doesn't represent the amount of oxygen dissolved in the water, it's the amount of oxygen in the air that's still left in the hull (or a value that doesn't affect anything if the hull is full of water). I would find it very strange if the rising water level in a hull would make the air quality worse, and that you wouldn't be able to breathe if the water rises up to your shoulders because there's only x% oxygen in the remaining air. Or that a deep swimming pool would be lethal because the room volume is mostly taken up by water, magically making the air above it unbreathable.

**Reply by NotWendy** [reply]
Hm, I may be misunderstanding how breathing works in Barotrauma then. (Go figure, you're never low on oxygen in this game.)
I assumed that the room has a certain amount of oxygen that can fit in it, and as long as there's *some* oxygen in the air, you won't pass out. And the vents deliver oxygen into the room at a certain rate, based on the room's full volume.
So I assumed that with, say, 80% of the room being filled with water, the room wouldn't be able to hold as much oxygen in it, but those 20% which would be available would still be enough for characters to breathe.
If that's not the case, then it makes my suggestion a lot harder to implement.

Basically, my assumption was that the amount of UNITS of oxygen in the air and the % of air quality in the room are not the same thing. One is the concrete amount of breathable oxygen, which the character consumes. The other is just the proportion between how many units of oxygen are in the room and how many units of oxygen the room can hold in total, based on its FULL volume.

Or maybe that is how it works and it's my suggestion that isn't phrased clearly. The idea is that water effectively reduces the volume of the room (without affecting the rate at which vents pump oxygen into said room). And a smaller dry space can hold less oxygen in it. The excess oxygen disappears.

**Reply by Regalis11** [reply]
Ah, I misunderstood you then. I thought you actually meant that a 95% flooded room would mean the remaining amount of air could only have 5% oxygen in it (or, have such a low relative amount of oxygen that you'd suffocate). And to be clear, when we're talking about percentages of oxygen, 100% in the game clearly doesn't mean you're breathing pure oxygen (which would be lethal), but rather that the air quality is "as good as it can be".

I think one thing we need to consider here, and something that's kind of just handwaved away by the game, is where the air in the sub comes from. When you're emptying a flooded room, you're clearly not forming a vacuum that must be filled by the oxygen generator - there's air coming into the room from _somewhere_. Afaik real submarines use high-pressure air tanks to fill the ballast tanks when they're pumping out water, perhaps we can imagine the subs in Barotrauma work the same way and that there's some sort of an air reservoir that flooding pushes the air into, and from which the air comes back into the hull as the water is pumped out? :barodev:

If we did switch to more realistic modeling of the air and oxygen, I think we would need to give an actual explanation to where the air comes from, and how emptying a fully flooded sub doesn't turn it into a vacuum chamber.

**Reply by NotWendy** [reply]
You're right, yeah. I was thinking about it purely from the gameplay perspective, but the air itself does need to come from somewhere. Oxygen generator only replenishes oxygen, as the name suggests. The main thing you'd need to replenish if a flood pushes air out of your submarine is nitrogen. And you'd need a lot of it.

**Reply by ubertpendragon** [reply]
Oxygen generation is a non-factor right now, having flooded hulls remove oxygen (instantly or over time) would at least be a little more interesting

**Reply by NotWendy** [reply]
Yeah, it's just a matter of how that mechanic can be framed to make sense lore-wise. I'm sure there are ways. Barotrauma's world doesn't need more inconsistencies, you're already fighting an uphill battle as a player, trying to suspend your disbelief.
